gcs-action/src/gcs.ts

```diff
@@ -0,0 +1,202 @@
+  }
+}
+
+/**
+ * List all files in a bucket, filtered by prefix
+ * based on https://github.com/googleapis/nodejs-storage/blob/master/samples/listFilesByPrefix.js
+ * @param bucketName The name of the bucket
+ * @param prefix The prefix to filter by
+ * @param delimiter Optional delimiter
+ */
+export async function listFilesByPrefix(
+  bucketName: string,
+  prefix: string,
+  delimiter?: string
+): Promise<File[]> {
+  //interface for listing options
+  interface ListOptions {
+    prefix: string
+    delimiter?: string
+  }
+  const options: ListOptions = {prefix}
+
+  if (delimiter) {
+    options.delimiter = delimiter
+  }
+  // Lists files in the bucket, filtered by a prefix
+  const [files] = await storage.bucket(bucketName).getFiles(options)
+  return files
+}
+
+/**
+ * Deletes a file in a bucket
+ * based on https://github.com/googleapis/nodejs-storage/blob/master/samples/deleteFile.js
+ * @param bucketName The name of the bucket
+ * @param filename The name of the name of the file to be deleted
+ */
+export async function deleteFile(
+  bucketName: string,
+  filename: string
+): Promise<{status: DeleteFileResponse}> {
+  try {
+    const deletedFile = await storage
+      .bucket(bucketName)
+      .file(filename)
+      .delete()
+    console.log(`gs://${bucketName}/${filename} deleted.`)
+    return {status: deletedFile}
+  } catch (err) {
+    return {status: err}
+  }
+}
+
+/**
+ * Uploads a specified directory to a GCS bucket
+ * based on https://github.com/googleapis/nodejs-storage/blob/master/samples/uploadDirectory.js
+ * @param bucketName The name of the bucket
+ * @param directoryPath The path of the directory to upload
+ * @param objectKeyPrefix Optional Prefix for in the GCS bucket
+ * @param clearExistingFilesFirst Clean files in the prefix before uploading
+ */
+export async function uploadDirectory(
+  bucketName: string,
+  directoryPath: string,
+  objectKeyPrefix: string = '',
+  clearExistingFilesFirst: boolean = false
+): Promise<{fileName: string; status: UploadResponse | Error}[] | undefined> {
+  // to store list of all files
+
+  const fileList: string[] = []
+  // get the list of files from the specified directory
+  let dirCtr = 1
+  let itemCtr = 0
+  const pathDirName = path.dirname(directoryPath)
+
+  /**
+   * Uploads a list of files to GCS
+   */
+  async function onComplete(): Promise<
+    {fileName: string; status: UploadResponse | Error}[]
+  > {
+    // if clearExistingFilesFirst, clear all files inside objectKeyPrefix
+    // if bucket contains /a/1.txt and /a/b/2.txt and prefix is /a
+    // /a/1.txt and /a/b/2.txt will both be cleared
+    if (clearExistingFilesFirst) {
+      try {
+        await listFilesByPrefix(bucketName, objectKeyPrefix).then(files => {
+          files.map((file: {name: string}) => {
+            deleteFile(bucketName, file.name)
+          })
+        })
+      } catch (err) {
+        core.setFailed(`Action failed with error ${err}`)
+        //console.log(err)
```

Comment:
What behavior are you trying to achieve here?

Reply:
My initial thoughts when approaching this action was to capture the files that errors instead of letting it bubble up. I think its cleaner to just let it bubble up like you suggested and set error in main